server/src/main/scala/services/crunch/deskrecs/TerminalDeskRecs.scala

```diff
@@ -0,0 +1,123 @@
+    .toMap
+
+  def adjustedWork(queue: Queue, work: Seq[Double]): Seq[Double] = queue match {
+    case EGate => work.map(_ / bankSize)
+    case _ => work
+  }
+
+  def minMaxDesksForQueue(deskRecMinutes: Iterable[MillisSinceEpoch], tn: Terminal, qn: Queue): (List[Int], List[Int]) = {
+    val defaultMinMaxDesks = (List.fill(24)(0), List.fill(24)(10))
+    val queueMinMaxDesks = minMaxDesks.getOrElse(tn, Map()).getOrElse(qn, defaultMinMaxDesks)
+    val minDesks = deskRecMinutes.map(desksForHourOfDayInUKLocalTime(_, queueMinMaxDesks._1))
+    val maxDesks = deskRecMinutes.map(desksForHourOfDayInUKLocalTime(_, queueMinMaxDesks._2))
+    (minDesks.toList, maxDesks.toList)
+  }
+
+  def terminalWorkToDeskRecs(terminal: Terminal,
+                             minuteMillis: NumericRange[MillisSinceEpoch],
+                             terminalPax: Map[Queue, Seq[Double]],
+                             terminalWork: Map[Queue, Seq[Double]],
+                             terminalRecs: TerminalDeskRecsProviderLike): Iterable[DeskRecMinute] = {
+    val terminalMinMaxDesks = queuesByTerminal(terminal).map { queue =>
+      (queue, minMaxDesksForQueue(minuteMillis, terminal, queue))
+    }.toMap
+    val minDesks = terminalMinMaxDesks.mapValues(_._1)
+    val maxDesks = terminalMinMaxDesks.mapValues(_._2)
+
+    val queueDesksAndWaits = terminalRecs.desksAndWaits(terminalWork, minDesks, maxDesks)
+
+    queueDesksAndWaits.flatMap {
+      case (queue, (desks, waits)) =>
+        minuteMillis.zip(terminalPax(queue).zip(terminalWork(queue))).zip(desks.zip(waits)).map {
+          case ((minute, (pax, work)), (desk, wait)) => DeskRecMinute(terminal, queue, minute, pax, work, desk, wait)
+        }
+    }
+  }
+}
+
+case class StaticTerminalDeskRecsProvider(queuesByTerminal: SortedMap[Terminal, Seq[Queue]], minMaxDesks: Map[Terminal, Map[Queue, (List[Int], List[Int])]], slas: Map[Queue, Int], cruncher: TryCrunch, bankSize: Int) extends TerminalDeskRecsProviderLike
+
+case class FlexedTerminalDeskRecsProvider(queuesByTerminal: SortedMap[Terminal, Seq[Queue]], minMaxDesks: Map[Terminal, Map[Queue, (List[Int], List[Int])]], slas: Map[Queue, Int], terminalDesks: Int, flexedQueuesPriority: List[Queue], cruncher: TryCrunch, bankSize: Int) extends TerminalDeskRecsProviderLike {
+  override def desksAndWaits(loads: Map[Queue, Seq[Double]],
+                             minDesks: Map[Queue, List[Int]],
+                             maxDesks: Map[Queue, List[Int]]): Map[Queue, (List[Int], List[Int])] = {
+    val queuesToOptimise: Set[Queue] = loads.keys.toSet
+    val flexedQueuesToOptimise = queuesToOptimise.filter(q => flexedQueuesPriority.contains(q))
+    val staticQueuesToOptimise = queuesToOptimise.filter(q => !flexedQueuesPriority.contains(q))
+
+    val flexedRecs = flexedDesksAndWaits(flexedQueuesToOptimise, loads, minDesks)
+
+    val staticRecs = super.desksAndWaits(loads.filterKeys(staticQueuesToOptimise), minDesks, maxDesks)
```

Comment:
you could actually rename the super.desksAndWaits to staticDesksAndWaits, then you'd be able to call if from both classes, I think it would make it clearer how things work.

Reply:
good call. done